# Week 08 - Teng Yong Khoo 

# Personal data exploration of literate Jupyter Notebook on the data for the three people agreed within the group

# The total time taken to complete this checkpoint is 8 hours 19 minutes.

## Section 1: Gaining Familiarity with our data

### Background & Purpose of the Analysis

Heart disease is the number one killer in Australia for decades. One of the many reasons why heart attack occurs is a lack of physical activity. Being inactive for a substantial period of time can cause fatty material building up in the individual's arteries. If these arteries which carries blood to the heart get clogged, it can lead to a heart attack which could have fatal consequences. 

In this analysis, we will be exploring on whether Fitbit users that are included in the dataset avoid inactivity in at least 10 hours a day. 

(Beginning of Work: 13th September 2024, End of Work: 13th September 2024)

### Loading the Packages

In [1]:
# Importing the packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# (Beginning of Work: 13th September 2024, End of Work: 13th September 2024)

### Loading the Dataset

Let us load the packages that are neccessary to conduct the analysis!

Now, let us load the 2 datasets using the pandas read_csv function that contains the dataset the shows the step count per day and per hour for each individual:

This dataset displays the step count per day for each individual:

(Beginning of Work: 13th September 2024, End of Work: 13th September 2024)

In [2]:
# Loading the dataset
df_day = pd.read_csv("dailySteps_merged.csv")

# Displaying the dataset
df_day

# (Beginning of Work: 13th September 2024, End of Work: 13th September 2024)

Id ActivityDay  StepTotal
0    1503960366   4/12/2016      13162
1    1503960366   4/13/2016      10735
2    1503960366   4/14/2016      10460
3    1503960366   4/15/2016       9762
4    1503960366   4/16/2016      12669
..          ...         ...        ...
935  8877689391    5/8/2016      10686
936  8877689391    5/9/2016      20226
937  8877689391   5/10/2016      10733
938  8877689391   5/11/2016      21420
939  8877689391   5/12/2016       8064

[940 rows x 3 columns]

This dataset displays the step count per hour for each individual:

In [3]:
# Loading the dataset
df_min = pd.read_csv("minuteStepsWide_merged.csv")

# Displaying the dataset
df_min

# (Beginning of Work: 13th September 2024, End of Work: 13th September 2024)

Id           ActivityHour  Steps00  Steps01  Steps02  Steps03  \
0      1503960366  4/13/2016 12:00:00 AM        4       16        0        0   
1      1503960366   4/13/2016 1:00:00 AM        0        0        0        0   
2      1503960366   4/13/2016 2:00:00 AM        0        0        0        0   
3      1503960366   4/13/2016 3:00:00 AM        0        0        0        0   
4      1503960366   4/13/2016 4:00:00 AM        0        0        0        0   
...           ...                    ...      ...      ...      ...      ...   
21640  8877689391   5/13/2016 3:00:00 AM        0        0        0        0   
21641  8877689391   5/13/2016 4:00:00 AM        0        0        0        0   
21642  8877689391   5/13/2016 5:00:00 AM        0        0        0        0   
21643  8877689391   5/13/2016 6:00:00 AM        0        0        0        0   
21644  8877689391   5/13/2016 7:00:00 AM       35       21       44        0   

       Steps04  Steps05  Steps06  Steps07  ...  Steps50  Steps51  Steps52  \
0            0        9        0       17  ...        0        9        8   
1            0        0        0        0  ...        0        0        0   
2            0        0        0        0  ...        0        0        0   
3            0        0        0        0  ...        0        0        0   
4            0        0        0        0  ...        0        0        0   
...        ...      ...      ...      ...  ...      ...      ...      ...   
21640        0        0        0        0  ...        0        0        0   
21641        0        0        0        0  ...        0        0        0   
21642        0        0        0        0  ...        0        0        0   
21643        0        0        0        0  ...       21       22       33   
21644        6        0        0        0  ...        0        0        0   

       Steps53  Steps54  Steps55  Steps56  Steps57  Steps58  Steps59  
0            0       20        1        0        0        0        0  
1            0        0        0        0        0        0        0  
2            0        0        0        0        0        0        0  
3            0        0        0        0        0        0        0  
4            0        0        0        0        0        0        0  
...        ...      ...      ...      ...      ...      ...      ...  
21640        0        0        0        0        0        0        0  
21641        0        0        0        0        0        0        0  
21642        0        0        0        0        0        0        0  
21643       17       24        0        0       20       38       16  
21644        0        0        0        0        0        0        0  

[21645 rows x 62 columns]

We are specifically interested in knowing the summary statistics of just 3 individuals. Note in both datasets, the user IDs are being repeated for each individual based on the day and minute. 

The Python function unique() will filter out the duplicates and extract just one unique user ID for each datapoint in ActivityDay and ActivityHour. For this analysis, we will be extracting users from rows 7 to 9. 

(Beginning of Work: 13th September 2024, End of Work: 13th September 2024)

In [4]:
# Filtering duplicate user IDs using unique()
indexday = df_day["Id"].unique()

# Printing the ID
print("The User ID is:", indexday[6], indexday[7], indexday[8])

# Filtering duplicate user IDs using unique()
indexminute = df_min["Id"].unique()

# Printing the ID
print("The User ID is:", indexminute[6], indexminute[7], indexminute[8])

# (Beginning of Work: 13th September 2024, End of Work: 13th September 2024)

The User ID is: 2026352035 2320127002 2347167796
The User ID is: 2026352035 2320127002 2347167796


As you can see from the above, the user ID for both datasets appears to be similar, which tells us the unique() function has done its part to only select one unique user ID for each datapoint in ActivityDay and ActivityHour.

(Beginning of Work: 13th September 2024, End of Work: 13th September 2024)

### Initial Assumptions
- There are no NAN for both datasets.

- The number of days of data for both datasets for each unique ID is the same. 

**These assumptions are essential, so that the final results would not be biased and inconsistent.**

(Beginning of Work: 13th September 2024, End of Work: 16th September 2024)

### Initial Predictions

As there are a substantial number of zeros in both datasets, it might not capture the true number of steps taken for each of the 3 individuals, and the steps count would be lower than expected.

(Beginning of Work: 16th September 2024, End of Work: 16th September 2024)

# Section 2: Exploratory Data Analysis (EDA)

### Data Cleaning

Before we begin our EDA, it is crucial to ensure that both datasets do not contain any outliers or NAN values which can affect the overall result for the analysis. 

First, we will be using the isna().sum() function to detect whether there are any NAN values in for the daily step count dataset for column "ActivityDay". As for the steps per minute dataset, we will use the function isna().sum().sum() for such wide dataset. The additional sum() at the end of the first sim() adds up the total number of missing values for each minute of step count. 

We will also be counting how many data points are recorded as zeros for the daily step count dataset. A calculation of the number of zero data points for the minute data set is not needed as it is crystal clear for the dataset that is loaded above, a substantial number of zero data points are present. 

(Beginning of Work: 13th September 2024, End of Work: 13th September 2024)

In [5]:
# Using isna().sum to detect if there's any NAN values in the ActivityDay column
nanday = df_day["ActivityDay"].isna().sum()

# Print the result
nanday

# (Beginning of Work: 13th September 2024, End of Work: 13th September 2024)

0

In [6]:
# Using isna().sum().sum() to detect if there's any NAN values in the entire dataset 
nanmin = df_min.isna().sum().sum()

# Print the result
nanmin

# (Beginning of Work: 13th September 2024, End of Work: 13th September 2024)

0

In [29]:
# Counting the number of zeros data points for the daily step count dataset. 
zero = (df_day['StepTotal'] == 0).sum()

# Print the result
zero

# (Beginning of Work: 16th September 2024, End of Work: 16th September 2024)

77

It seems like there are no missing values for both datasets, which means the assumption of no NAN values is fulfilled. 

There are 77 zero data points in the daily step count dataset, as well as a visual observation from the minute step count dataset which shows multiple zero data points. This may indicate that users may have forgotten to wear their tracking devices, or did not turn it on when walking, or they simply just do not walk.

(Beginning of Work: 13th September 2024, End of Work: 16th September 2024)

### Number of Days of Data - Daily Step Count Dataset

We are now ready to perform our EDA. First, we will display the number of days of data for each of the 3 individuals.

We will first use the pandas "to_datetime" function to convert the column ActivityDay column to a datetime format. This operation because it provides a more accurate counting of the days of the data for each of the 3 individuals. We will then create a list to indicate which are the 3 users that we want to analyse, then use the isin() function to filter the dataframe such that it only contains information for these 3 individuals.

To avoid duplicates, we will be using the function .nunique() to indicate to Python to only count the unique number of days, meaning if there are multiple same dates for the particular user, Python will only count as 1 day. 

(Beginning of Work: 15th September 2024, End of Work: 16th September 2024)

In [7]:
# Convert 'ActivityDay' to datetime format
df_day['ActivityDay'] = pd.to_datetime(df_day['ActivityDay'], format='%m/%d/%Y')

# The 3 user IDs to analyze
userid = [2026352035, 2320127002, 2347167796]

# Filter the dataframe to only include the 3 user IDs by creating a new variable dfnew1
dfnew1 = df_day[df_day['Id'].isin(userid)]

# Count the number of unique days for each user ID using .nunique()
result = dfnew1.groupby('Id')['ActivityDay'].nunique()

# Printing the results 
print(result)

# (Beginning of Work: 15th September 2024, End of Work: 16th September 2024)

Id
2026352035    31
2320127002    31
2347167796    18
Name: ActivityDay, dtype: int64


### Number of Days of Data - Minute Step Count Dataset

Since the minute step count dataset is in a wide format where there are multiple dates that are similar for each row, we need to perform one addtional step, which is to create an additional column "ActivityDate" such that all the information about the number of minutes of step count for each individual are aggregated into one single date. 

The purpose of the function dt.date is to extract the date part of the datetime.

(Beginning of Work: 15th September 2024, End of Work: 15th September 2024)

In [ ]:
# First, extract the date from the 'ActivityHour' column
df_min['ActivityDate'] = pd.to_datetime(df_min['ActivityHour']).dt.date

# Filter DataFrame for the specified User IDs
dfnew2 = df_min[df_min['Id'].isin(userid)]

result2 = dfnew2.groupby('Id')['ActivityDate'].nunique()

# Printing the result
print(result2)

# (Beginning of Work: 15th September 2024, End of Work: 16th September 2024)

For the daily step count dataset, user ID 2026352035 and 2320127002 have a total of 31 days of data. While for user ID 2347167796, there are 18 days of data. 

As for the minute step count dataset, user ID 2026352035 and 2320127002 have a total of 31 days of data. While for user ID 2347167796, there are only 17 days of data, one short of the daily step count dataset.

**Note: This is contrary to the initial assumption where we assume that the number of days of data are the same for each user IDs for both datasets.**

(Beginning of Work: 16th September 2024, End of Work: 16th September 2024)

### Daily Step Count Information

We are interested in knowing the daily step count information for each of the 3 individuals. To do this we will be creating a function using "def" to avoid any repetitive code. For the following code, we will be defining our function as "dailyinfo". This function takes in two arguments, df_day and userid. "df_day" is a dataframe that contains information within the daily step count dataset, while "userid" compiles the list of the users that will be used for this analysis.

The function isin() would filter our analysis for the 3 individuals only as mentioned earlier. We will then get the average step count per day using the function mean(), the maximum step count using max(), the minimum step count using min(), and one additional observation which is the standard deviation using std(). 

We will then use 'return' to tell Python to send back the results to the function that was defined earlier. 

(Beginning of Work: 16th September 2024, End of Work: 16th September 2024)

In [17]:
# Defining a function that takes in the 2 arguments: "df_day" & "userid"
def dailyinfo(df_day, userid):
    # Using the isin() function to filter users within the list that was mentioned earlier
    dfnew1 = df_day[df_day['Id'].isin(userid)] 
    
    averageday = dfnew1.groupby('Id')['StepTotal'].mean() # Average step counts per day
    maxday = dfnew1.groupby('Id')['StepTotal'].max() # Maximum step count
    minday = dfnew1.groupby('Id')['StepTotal'].min() # Minimum step count
    stdday = dfnew1.groupby('Id')['StepTotal'].std() # Standard Deviation 
    
    # Sending back the results to the function 
    return averageday, maxday, minday, stdday

# (Beginning of Work: 16th September 2024, End of Work: 16th September 2024)

In [18]:
# Receiving the results
averageday, maxday, minday, stdday = dailyinfo(df_day, userid)

# Printing the information
print("Average Step Count per Day:\n", averageday)

print("Maximum Step Count per Day:\n", maxday)

print("Minimum Step Count per Day:\n", minday)

print("Standard Deviation of Step Count:\n", stdday)

# (Beginning of Work: 16th September 2024, End of Work: 16th September 2024)

Average Step Count per Day:
 Id
2026352035    5566.870968
2320127002    4716.870968
2347167796    9519.666667
Name: StepTotal, dtype: float64
Maximum Step Count per Day:
 Id
2026352035    12357
2320127002    10725
2347167796    22244
Name: StepTotal, dtype: int64
Minimum Step Count per Day:
 Id
2026352035    254
2320127002    772
2347167796     42
Name: StepTotal, dtype: int64
Standard Deviation of Step Count:
 Id
2026352035    2977.834815
2320127002    2254.614642
2347167796    4682.093700
Name: StepTotal, dtype: float64


The summary statistics for each user ID are as follows:

**User ID: 2026352035**

- Average Step Count Per Day: Approximately 5567 steps
- Maximum Step Count: 12357 steps
- Minimum Step Count: 254 steps
- Standard Deviation of Step Count: 2977.83

Observation: It looks like this person is fairly active with a average step count of just over 5000 steps, with their most active day recorded with 12357 steps. 

**User ID: 2320127002**

- Average Step Count Per Day: Approximately 4717 steps
- Maximum Step Count: 10725 steps
- Minimum Step Count: 772 steps
- Standard Deviation of Step Count: 2254.61

Observation: This person could be more active as their average step count is just below 5000, at 4717 steps. 

**User ID: 2347167796**

- Average Step Count Per Day: Approximately 9520 steps
- Maximum Step Count: 22244 steps
- Minimum Step Count: 42 steps
- Standard Deviation of Step Count: 4682.09

Observation: This person is very active with an average step count per day of 9520 steps, which is just a few hundred steps short of 10,000 steps as recommended by many health experts. The standard deviation for user ID: 2347167796 is high, which suggests that it fluctuates quite a lot compared to other 2 individuals which have lesser variation. 

(Beginning of Work: 16th September 2024, End of Work: 16th September 2024)

### Minute Step Count Information

Previously, we have converted our the ActivityDay column into datetime for a more accurate counting of the number of days in the dataset. We now need to reverse this action as Python is unable to compare things like > 0 between datetime values and integers. We will then reload the minute step count dataset and define a new variable df_min2. 

The process to obtain information of the average steps per minute, and the maximum and minimum steps are similar to the daily step count dataset, with the exception of that the commands such as .mean() are repeated again to use all the information for each minute in each column to calculate the relevant statistics. 

As the relevant data starts from column 3, we will use the function iloc[:, 2:] to inidicate that we want to generate the statistics from column 3 onwards. We will then return the function for each user individually and print the results in 3 seperate code blocks. 

The one additional observation that I've made for each person is the total number of minutes that the user is active. This figure will give me a preliminary insight on how active each of the 3 individuals are. 

(Beginning of Work: 16th September 2024, End of Work: 16th September 2024)

In [52]:
# Loading the dataset
df_min2 = pd.read_csv("minuteStepsWide_merged.csv")

# Displaying the dataset
df_min2

# Defining the function 
def minuteinfo(df_min2, user_id):
    dfnew2 = df_min2[df_min2['Id'] == user_id]
    
    nonzero = (dfnew2.iloc[:, 2:] > 0).sum().sum() # Number of non-zero minutes
    averageminute = dfnew2.iloc[:, 2:].mean().mean() # Average steps per minute 
    maxminute = dfnew2.iloc[:, 2:].max().max() # Maximum number of steps
    minminute = dfnew2.iloc[:, 2:].min().min() # Minimum number of steps
    missing = dfnew2.isna().sum().sum() # The number of missing data
    totalminute = (dfnew2.iloc[:, 2:] > 0).sum().sum() # The total number of minutes that the user is active
    
    # Returning the results to the function
    return nonzero, averageminute, maxminute, minminute, missing, totalminute

# Returning the results for User ID: 2026352035
nonzero1, averageminute1, maxminute1, minminute1, missing1, totalminute1 = minuteinfo(df_min2, 2026352035)

print("User ID: 2026352035\n")
print("Non-zero minutes:", nonzero1)
print("Average steps per minute:", averageminute1)
print("Maximum Step Count Per Minute:", maxminute1)
print("Minimum Step Count Per Minute:", minminute1)
print("Missing Data Points:", missing1)
print("Total active minutes:", totalminute1)

# (Beginning of Work: 16th September 2024, End of Work: 16th September 2024)

User ID: 2026352035

Non-zero minutes: 8022
Average steps per minute: 3.9721307727480566
Maximum Step Count Per Minute: 125
Minimum Step Count Per Minute: 0
Missing Data Points: 0
Total active minutes: 8022


In [54]:
# Returning the results for User ID: 2320127002
nonzero2, averageminute2, maxminute2, minminute2, missing2, totalminute2 = minuteinfo(df_min2, 2320127002)

print("User ID: 2320127002\n")
print("Non-zero minutes:", nonzero2)
print("Average steps per minute:", averageminute2)
print("Maximum Step Count Per Minute:", maxminute2)
print("Minimum Step Count Per Minute:", minminute2)
print("Missing Data Points:", missing2)
print("Total active minutes:", totalminute2)

# (Beginning of Work: 16th September 2024, End of Work: 16th September 2024)

User ID: 2320127002

Non-zero minutes: 6079
Average steps per minute: 3.1616140832190216
Maximum Step Count Per Minute: 123
Minimum Step Count Per Minute: 0
Missing Data Points: 0
Total active minutes: 6079


In [55]:
# Returning the results for User ID: 2347167796
nonzero3, averageminute3, maxminute3, minminute3, missing3, totalminute3 = minuteinfo(df_min2, 2347167796)

print("User ID: 2347167796\n")
print("Non-zero minutes:", nonzero3)
print("Average steps per minute:", averageminute3)
print("Maximum Step Count Per Minute:", maxminute3)
print("Minimum Step Count Per Minute:", minminute3)
print("Missing Data Points:", missing3)
print("Total active minutes:", totalminute3)

# (Beginning of Work: 16th September 2024, End of Work: 16th September 2024)

User ID: 2347167796

Non-zero minutes: 4580
Average steps per minute: 6.889914529914529
Maximum Step Count Per Minute: 167
Minimum Step Count Per Minute: 0
Missing Data Points: 0
Total active minutes: 4580


Observation: 

User ID 2347167796 seems to be the most active as analysed from the minute step count dataset where they have recorded approximately 7 steps per minute. This finding is consistent with the daily step count data. Moreover, there are no missing data points which is also consistent with our previous finding.

It's interesting to see that while user ID 2347167796 does not have the highest number of minutes active (in fact the lowest). However, they have the highest average steps per minute which is a far more accurate representation of being active. 

(Beginning of Work: 16th September 2024, End of Work: 16th September 2024)

### Final Statement & Relationship to the Driving Problem

From the data exploration, I've learnt that the majority of the users here are not really walking actively as their step counts are significantly below the recommended number of steps to walk per day by health experts. However, it is important to note that with such a small sample size, it is naive to deduce the fact that the users in the entire dataset are not actively walking.

The analysis performed above relates to the driving problem well, especially the part where zero values and NAN values are being checked. Having zero values in the dataset is something that we need to be aware of. Sometimes it may indicate the user might forgot the wear their devices when walking, but there are also times where the user simply do not walk. This is a crucial finding for further stages of the analysis, and this is something that needs to be taken into account when conducting further analysis. 

Finally, while user ID no. 2347167796 seems to be most active according to the summary statistics, it is not possible to conclude at this stage that this user is an active individual, due to the earlier assumption violation. This problem will need to be addressed at later stages. 

(Beginning of Work: 16th September 2024, End of Work: 16th September 2024)

## References

Alexander, T. 2024. COSC2902: Computational Modelling (Advanced). The University of Sydney
Available at: https://canvas.sydney.edu.au/courses/56907/pages/week-2-simulating-randomness?module_item_id=2196702

Poon, J. 2023. DATA1902: Informatics: Data and Computation (Advanced). The University of Sydney
Available at: https://edstem.org/au/courses/12221/lessons/38247/slides/265109

(Beginning of Work: 16th September 2024, End of Work: 16th September 2024)

## Acknowledgement of the use of AI

The code was generated using generative AI (GAI), specifically Open AI ChatGPT. The prompts used and the the time of content generated are as follows:

- "Provide one other information other than the mean, maximum and minimum step count for the steps per day dataset" (Generated at: 2:24pm, 16th September 2024).


- "Provide one other information other than the mean, maximum and minimum step count, non-zero minutes, and missing data for the steps per minute dataset" (Generated at: 4:08pm, 16th September 2024). 


- "How should I filter the users defined above as "userid = [2026352035, 2320127002, 2347167796]". (Generated at: 9:34pm, 15th September 2024). 


- "How should I generate the mean, maximum and minimum step count and non-zero minutes for the steps per minute dataset for each user ID". (Generated at: 2:56pm, 16th September 2024). 


There was a mixutre of codes that was used directly, that is for the first 2 commands. While for the last 2 commands, some parts of codes generated by AI were used, with some modifications made using concepts from DATA1902, and COSC2902. Specifically, the parts of code used from AI is where I applied used the function "iloc", and also the part where I want to generate the statistics by user ID.

(Beginning of Work: 16th September 2024, End of Work: 16th September 2024)